### PR TITLE
Allow multiple alternative aliases

### LINKED ISSUE #25: Discussion: Allow multiple aliases?
At GlobalWebIndex we have our own custom [`ConsistentImports` rule](https://gist.github.com/Janiczek/96fda8308a5e0bc0bee1fbe37aee4214).

It allows multiple aliases (we need this because of us having a convention `import List.Extra as List`, and `Html.Attributes.Extra.autocomplete` colliding with `Html.Attributes.autocomplete`) - described more in the [Elm Slack discussion](https://elmlang.slack.com/archives/C010RT4D1PT/p1592564858278800).

We wanted to publish our rule, then found out this package exists, so we'd love to not divide the ecosystem.
Would you consider allowing multiple aliases for a given imported module in `elm-review-imports`?

### PR DESCRIPTION
**Pull request by sparksp**
Adds an additional list to the config tuple (so that at least one alias must be provided). Fixes #25 

The rule will...

- [x] Enforce the first unique alias from the config.
- [x] If there are no unique aliases available use the **last** configured alias.
- [x] Only offer fixes when the change is to **and** from a unique alias.
    - When changing from a shared alias we don't know which alias uses to change (and we're not going to look it up).
    - When changing to a shared alias we could introduce a compiler error if both modules use the same function (and we're not going to check).
- [x] Check for alias availability in the order the modules are **config**'d.

## Success

```elm
config : List Rule
config =
    [ NoInconsistentAliases.config
        [ ( "Html.Attributes", "Attr", [] )
        , ( "My.Html.Attributes", "Attr", [] )
        , ( "Html.Attributes.Extra", "Attr", [ "Attr_" ] )
        , ( "Svg.Attributes", "Attr", [ "SvgAttr" ] )
        , ( "My.Svg.Attributes", "Attr", [ "SvgAttr" ] )
        ]
        |> NoInconsistentAliases.noMissingAliases
        |> NoInconsistentAliases.rule
    ]

--- (1)
import Html.Attributes as Attr
import Html.Attributes.Extra as Attr_
import My.Html.Attributes as Attr
import My.Svg.Attributes as SvgAttr
import Svg.Attributes as SvgAttr

--- (2)
import Html.Attributes.Extra as Attr
import My.Svg.Attributes as SvgAttr
import Svg.Attributes as SvgAttr

--- (3)
import My.Svg.Attributes as SvgAttr
import Svg.Attributes as Attr
```

### Notes (1)

`Html.Attributes` and `My.Html.Attributes` use their only allowed alias (`Attr`), `Html.Attributes.Extra` uses it's fallback alias (`Attr_`), and `Svg.Attributes` and `My.Svg.Attributes` use their fallback alias (`SvgAttr`).

### Notes (2)

`Html.Attributes.Extra` is configured first so gets `Attr` - Svg modules then use their fallback.

### Notes (3)

`Svg.Attributes` is configured first so gets `Attr`, `My.Svg.Attributes` uses its fallback.  Without a grouping system there's no way to configure that these modules should use the same alias.

### CODE REVIEW

**Comment by sparksp** [comment]
Hey @Janiczek, I think this is complete now - would you mind casting an eye over it to see if it meets your needs and expectations?

File: src/NoInconsistentAliases/Config.elm [hunk 1]
```diff
@@ -1,33 +1,38 @@
...
     ( Config, config, noMissingAliases
-    , canMissAliases, lookupAlias
+    , canMissAliases, listAliases
     )
 
 {-|
 
 @docs Config, config, noMissingAliases
-@docs canMissAliases, lookupAlias
+@docs canMissAliases, lookupAliases, listAliases
```
Comment (Janiczek) [inline src/NoInconsistentAliases/Config.elm:9]:
```suggestion
@docs canMissAliases, listAliases
```

**Comment by Janiczek** [comment]
@sparksp I'll try using it instead of our current rule, and will get back to you with feedback! :+1:

**Comment by Janiczek** [comment]
So as it currently stands, there is a (to me) weird behaviour of the fallback aliases.

Config contains:
```elm
        , ( "Html.Attributes", "Attrs", [] )
        , ( "Html.Attributes.Extra", "Attrs", [ "Attrs_" ] )
```
And I'd expect it to be OK with both of the modules *primarily* using `Attrs` (if compiler doesn't complain about ambiguity of some used var like that `autocomplete` one).

But for code:
```elm
import Html.Attributes as Attrs
import Html.Attributes.Extra as Attrs
```
it raises an error:
```
NoInconsistentAliases: Incorrect alias `Attrs` for module
`Html.Attributes.Extra`.

35| import Html.Attributes as Attrs
36| import Html.Attributes.Extra as Attrs
                                    ^^^^^
37| import Html.Events as Events

This import does not use your preferred alias `Attrs_` for
`Html.Attributes.Extra`.

Your preferred alias has already been used by another module so you should
review carefully whether to overload this alias or configure another.

If you change this alias remember to change all references to the alias in this
module too.
```
So... am I understanding the roles of the aliases in the nonempty list wrong?

**Comment by sparksp** [comment]
This was one of the questions in the [discussion](https://github.com/sparksp/elm-review-imports/issues/25#issuecomment-651908402):

> > (4) Would you expect `import Html.Attributes.Extra as Attrs` to be reported if something else is also aliased to `Attrs` (because there's another alias it could use)? Or would it be ignored because it's using one of it's allowed aliases?
>
> Yes, that would be nice.

So we have two conflicting requirements at the moment, and I'm not sure how to describe these in code.

## Imports that can use any of the configured aliases

```elm
NoInconsistentAliases.config
    [ ( "Html.Attributes", "Attrs", [] )
    , ( "Html.Attributes.Extra", "Attrs", [ "Attrs_" ] )
    ]

--- PASS
import Html.Attributes as Attrs
import Html.Attributes.Extra as Attrs

--- PASS
import Html.Attributes as Attrs
import Html.Attributes.Extra as Attrs_
```

## Imports that must not conflict with other aliases

```elm
NoInconsistentAliases.config
    [ ( "Html.Attributes", "Attrs", [] )
    , ( "Svg.Attributes", "Attrs", [ "SvgAttrs" ] )
    ]

--- FAIL
import Html.Attributes as Attrs
import Svg.Attributes as Attrs

--- PASS
import Html.Attributes as Attrs
import Svg.Attributes as SvgAttrs

--- PASS
import Svg.Attributes as Attrs
```

**Comment by Janiczek** [comment]
Ah, sorry. I must have misread or misunderstood. Well, I fall into the first camp then: `Imports that can use any of the configured aliases`
But I understand how it can be a contentious point. :thinking:

**Comment by sparksp** [comment]
Okay, I think I'll close this PR and make a new one that just focusses on allowing any of the configured aliases. I don't want to lose the work done here because I think that it'll be useful when I figure out an API to configure "imports that must not conflict with other aliases".

**Pull request closed without merging**